text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+        Self {
+            // other fields...
+            _marker: PhantomData,
+        }
+    }
+}
+```
+
+Here, `_marker: PhantomData` is just provided to satisfy the compiler.
+`PhantomData` carries no useful information since it is a ZST.
+Therefore, the result of having to add a field of type `PhantomData`
+is a regression to ergonomics throughout. With default values you can
+improve the situation slightly and provide the value in a single place:
+
+```rust
+use std::marker::PhantomData;
+
+struct Vec<T> {
+    data: *const T, // *const for variance!
+    len: usize,
+    cap: usize,
+    _marker: PhantomData<T> = PhantomData,
+}
+```
+
+You then go on to define some operations for `Vec<T>`:
+
+```rust
+impl<T> Vec<T> {
+    pub fn new() -> Self {
+        Self {
+            // other fields...
+            ..
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            // other fields...
+            ..
+        }
+    }
+}
+```
+
+## `#[derive(Default)]` in more cases
+
+The `#[derive(..)]` ("custom derive") mechanism works by defining procedural
+*macros*. Because they are macros, these operate on abstract *syntax* and
+don't have more information available. Therefore, when you `#[derive(Default)]`
+on a data type definition as with:
+
+```rust
+#[derive(Default)]
+struct Foo {
+    bar: u8,
+    baz: String,
+}
+```
+
+it only has the immediate "textual" definition available to it.
+
+Because Rust currently does not have an in-language way to define default values,
+you cannot `#[derive(Default)]` in the cases where you are not happy with the
+natural default values that each field's type provides. By extending the syntax
+of Rust such that default values can be provided, `#[derive(Default)]` can be
+used in many more circumstances and thus boilerplate is further reduced.  The
+addition of a single field, expands the code written by the `struct` author from
+a single `derive` line to a whole `Default` `impl`, which becomes more verbose
+linearly with the number of fields.
+
+## Usage by other `#[derive(..)]` macros
+
+[`serde`]: https://serde.rs/attributes.html
+
+Custom derive macros exist that have a notion of or use default values.
+
+### `serde`
+
+For example, the [`serde`] crate provides a `#[serde(default)]` attribute that
+can be used on `struct`s, and fields. This will use the field's or type's
+`Default` implementations. This works well with field defaults; `serde` can
+either continue to rely on `Default` implementations in which case this RFC
+facilitates specification of field defaults; or it can directly use the default
+values provided in the type definition.
+
+### `structopt`
+
+Another example is the `structopt` crate with which you can write:
+
+```rust
+#[derive(Debug, StructOpt)]
+#[structopt(name = "example", about = "An example of StructOpt usage.")]
+struct Opt {
+    /// Set speed
+    #[structopt(short = "s", long = "speed", default_value = "42")]
```

Comment:
Nit: `#[structopt(default_value_t = 42)]` would be the more comparable structopt option.